### PR TITLE
Balder/all in single chunk

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR

### CODE REVIEW

**Review by vekterli — APPROVED** [review]
:+1: one more step towards less legacy dust hidden under the wooly document module carpet

File: document/src/vespa/document/serialization/vespadocumentserializer.cpp [hunk 1]
```diff
@@ -86,89 +86,45 @@ void VespaDocumentSerializer::write(const DocumentType &value) {
...
-        }
-    } else {
-        if (hasHeader) {
-            doc_serializer.write(value.getFields(), HeaderFields());
-        }
-        if (hasBody) {
-            doc_serializer.write(value.getFields(), BodyFields());
+    if ( hasContent ) {
+        if (!structNeedsReserialization(value.getFields())) {
+            doc_serializer.writeUnchanged(value.getFields().getFields());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/serialization/vespadocumentserializer.cpp]: That was at least the intention.

File: storage/src/tests/persistence/processalltest.cpp [hunk 2]
```diff
@@ -166,7 +165,7 @@ TEST_F(ProcessAllHandlerTest, stat_bucket_request_can_returned_removed_entries)
         "  Timestamp: 208, id:mail:testdoctype1:n=4:42967.html, gid(0x04000000f19ece1668e6de48) (remove)\n"
         "  Timestamp: 209, id:mail:testdoctype1:n=4:6925.html, gid(0x04000000667c0b3cada830be) (remove)\n";
 
-    EXPECT_EQ(expected, reply.getResults());
+        EXPECT_EQ(expected, reply.getResults());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/tests/persistence/processalltest.cpp]: Fixed

File: document/src/vespa/document/fieldvalue/serializablearray.cpp [hunk 3]
```diff
@@ -81,6 +81,13 @@ SerializableArray::SerializableArray(const SerializableArray& other)
     }
 }
 
+SerializableArray &
+SerializableArray::operator=(const SerializableArray &rhs)
+{
+    *this = SerializableArray(rhs);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/serializablearray.cpp]: Fixed

File: document/src/vespa/document/base/documentid.cpp [hunk 4]
```diff
@@ -29,7 +29,7 @@ DocumentId::DocumentId(vespalib::nbostream & is)
 
 DocumentId::DocumentId(const DocumentId & rhs) = default;
 DocumentId & DocumentId::operator = (const DocumentId & rhs) = default;
-DocumentId::~DocumentId() = default;
+DocumentId::~DocumentId() noexcept = default;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/base/documentid.cpp]: K, so more is better than less..

File: document/src/vespa/document/fieldvalue/serializablearray.cpp [hunk 5]
```diff
@@ -26,84 +26,103 @@ class BufferMap : public BufferMapT {
...
     for (size_t i(0); i < _entries.size(); i++) {
         Entry & e(_entries[i]);
         if (e.hasBuffer()) {
             // Pointing to a buffer in the _owned structure.
-            ByteBuffer::UP buf(ByteBuffer::copyBuffer(e.getBuffer(_uncompSerData.get()), e.size()));
-            e.setBuffer(buf->getBuffer());
-            ensure(_owned)[e.id()] = std::move(buf);
+            ByteBuffer buf(ByteBuffer::copyBuffer(e.getBuffer(&_uncompSerData), e.size()));
+            e.setBuffer(buf.getBuffer());
+            ensure(_unlikely->_owned)[e.id()] = std::move(buf);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/serializablearray.cpp]: No, that should not be possible.
e.hasBuffer will only be true if rhs._unlikely._owened has an entry for e._id.
And that again should guarantee that _unlikely is created in the initializer list.

File: document/src/vespa/document/base/idstring.h [hunk 6]
```diff
@@ -57,7 +59,7 @@ class IdString {
     private:
         static constexpr uint32_t MAX_COMPONENTS = 4;
         Offsets(vespalib::stringref id);
-        uint16_t _offsets[MAX_COMPONENTS + 1];
+        uint16_t _offsets[MAX_COMPONENTS];
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/base/idstring.h]: No, it did not change size due to alignment requirement. Just added padding.

File: document/src/vespa/document/fieldvalue/serializablearray.cpp [hunk 7]
```diff
@@ -26,84 +26,103 @@ class BufferMap : public BufferMapT {
...
 }
 
 void
-SerializableArray::set(int id, ByteBuffer::UP buffer)
+SerializableArray::set(int id, ByteBuffer buffer)
 {
     maybeDecompress();
-    Entry e(id, buffer->getRemaining(), buffer->getBuffer());
-    ensure(_owned)[id] = std::move(buffer);
+    Entry e(id, buffer.getRemaining(), buffer.getBuffer());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/serializablearray.cpp]: Fixed

**Pull request merged**